File: apps/ios/Pebbles/Features/PebbleMedia/SnapUploadCoordinator.swift
import Foundation
import os

/// Single source of truth for the snap (photo) attached to an in-progress
/// pebble form. Owns `formSnap` (the form-layer state), `processedForRetry`
/// (re-encoded bytes kept around for a user-tapped retry), and every
/// transition: pick → upload → retry → cancel → remove.
///
/// Both `CreatePebbleSheet` and `EditPebbleSheet` hold one of these and
/// delegate. All Storage side-effects are explicitly awaited — there is no
/// sleep-based race in `cancelAndCleanup`.
@MainActor
@Observable
final class SnapUploadCoordinator {

    // MARK: - State (publicly readable, mutated only by methods on this type)

    private(set) var formSnap: FormSnap?

    /// Processed bytes kept around so a `.failed → user-tapped retry` doesn't
    /// re-load + re-encode the original. Cleared whenever `formSnap` is
    /// cleared.
    private(set) var processedForRetry: ProcessedImage?

    // MARK: - Derived state

    /// True while the pending snap is still uploading.
    var isUploading: Bool {
        if case .pending(let snap) = formSnap, snap.state == .uploading { return true }
        return false
    }

    /// True when the pending snap's last upload attempt failed.
    var hasFailed: Bool {
        if case .pending(let snap) = formSnap, snap.state == .failed { return true }
        return false
    }

    /// `true` while save should be blocked because the snap is mid-upload
    /// or in a failed state requiring user action.
    var isBlocking: Bool { isUploading || hasFailed }

    /// Returns the pending `AttachedSnap` only when it has fully uploaded —
    /// suitable for embedding in a save payload. Returns `nil` for `.none`,
    /// `.existing` (those are encoded directly from `formSnap`), or any
    /// in-flight/failed `.pending` state.
    func pendingSnapForPayload() -> AttachedSnap? {
        if case .pending(let snap) = formSnap, snap.state == .uploaded { return snap }
        return nil
    }

    // MARK: - Init

    private let repo: PebbleSnapRepositoryProtocol
    private let retryDelay: Duration
    private static let logger = Logger(subsystem: "app.pbbls.ios", category: "snap-coordinator")

    init(
        repo: PebbleSnapRepositoryProtocol,
        initialSnap: FormSnap? = nil,
        retryDelay: Duration = .seconds(2)
    ) {
        self.repo = repo
        self.formSnap = initialSnap
        self.retryDelay = retryDelay
    }

    /// Set the initial snap *after* construction. Used by `EditPebbleSheet`
    /// which has to wait for `load()` to finish before it knows whether the
    /// pebble has an existing snap.
    func seedExisting(_ snap: FormSnap?) {
        self.formSnap = snap
    }

    // MARK: - Test seam

    #if DEBUG
    /// Test-only: seed `processedForRetry` directly so tests can exercise
    /// the upload path without driving the real `handlePicked` flow.
    func seedProcessedForRetry(_ processed: ProcessedImage) {
        self.processedForRetry = processed
    }
    #endif

    // MARK: - User-driven transitions

    func handlePicked(_ picked: PhotoPickerView.PickedItem, userId: UUID) async {
        Self.logger.notice("handlePicked: started uti=\(picked.uti, privacy: .public)")

        let data: Data
        do {
            data = try await Self.loadData(from: picked.itemProvider, uti: picked.uti)
            Self.logger.notice("handlePicked: loaded \(data.count, privacy: .public) bytes")
        } catch {
            Self.logger.error("picker data load failed: \(error.localizedDescription, privacy: .private)")
            return
        }

        let processed: ProcessedImage
        let uti = picked.uti
        do {
            processed = try await Task.detached(priority: .userInitiated) {
                try ImagePipeline.process(data, uti: uti)
            }.value
        } catch {
            Self.logger.error("image pipeline failed: \(String(describing: error), privacy: .public)")
            return
        }

        let snapId = UUID()
        formSnap = .pending(AttachedSnap(id: snapId, localThumb: processed.thumb, state: .uploading))
        processedForRetry = processed

        await performUpload(processed: processed, snapId: snapId, userId: userId)
    }

    /// Re-run the upload after a user tap on the chip's retry button. Only
    /// valid when the current state is `.pending(.failed)` and we still hold
    /// `processedForRetry`. No-op otherwise.
    func retryCurrent(userId: UUID) async {
        guard case .pending(var snap) = formSnap, let processed = processedForRetry else { return }
        snap.state = .uploading
        formSnap = .pending(snap)
        await performUpload(processed: processed, snapId: snap.id, userId: userId)
    }

    /// Remove the current `.pending` snap and fire the compensating Storage
    /// delete. Safe to call when the snap is in any state. No-op when the
    /// current `formSnap` is not `.pending`.
    func removePending(userId: UUID) async {
        guard case .pending(let snap) = formSnap else { return }
        let snapId = snap.id
        formSnap = nil
        processedForRetry = nil
        await repo.deleteFiles(snapId: snapId, userId: userId)
    }

    /// Remove the current `.existing` snap. Calls `delete_pebble_media` and
    /// then fires fire-and-forget Storage cleanup. Throws on RPC failure so
    /// the sheet can surface a user-facing error. No-op (returns normally)
    /// when the current `formSnap` is not `.existing`.
    func removeExisting() async throws {
        guard case .existing(let id, _) = formSnap else { return }
        let storagePath = try await repo.deletePebbleMedia(snapId: id)
        // Fire-and-forget Storage cleanup. The live repo logs on failure.
        await repo.deleteFiles(storagePrefix: storagePath)
        formSnap = nil
    }

    // MARK: - Sheet lifecycle

    /// Called by the sheet's Cancel button before `dismiss()`. Captures the
    /// current pending snap id, clears state, and *awaits* the compensating
    /// Storage delete. Storage-delete failures are logged but do not throw —
    /// Cancel must not be un-cancellable.
    func cancelAndCleanup(userId: UUID) async {
        // Capture id before clearing so the delete still has the address.
        guard case .pending(let snap) = formSnap else {
            formSnap = nil
            processedForRetry = nil
            return
        }
        let snapId = snap.id
        formSnap = nil
        processedForRetry = nil
        await repo.deleteFiles(snapId: snapId, userId: userId)
    }

    /// Called by the sheet after a save (compose-pebble / compose-pebble-update)
    /// failure when a pending snap is attached. Awaits the compensating
    /// Storage delete. Does *not* clear `formSnap` — the sheet stays open
    /// and the user can retry. Use `cancelAndCleanup` if you also want to
    /// dismiss.
    func handleSaveFailure(userId: UUID) async {
        guard case .pending(let snap) = formSnap else { return }
        await repo.deleteFiles(snapId: snap.id, userId: userId)
        // Intentionally do not clear formSnap — caller (the sheet) keeps the
        // form open so the user can retry.
    }

    // MARK: - Private

    /// Single upload path used by both `handlePicked` (initial attempt) and
    /// `retryCurrent` (user-tapped retry). On first failure, sleeps for
    /// `retryDelay` and retries once. On second failure, transitions the
    /// snap to `.failed` and retains `processedForRetry`.
    private func performUpload(processed: ProcessedImage, snapId: UUID, userId: UUID) async {
        do {
            try await repo.uploadProcessed(processed, snapId: snapId, userId: userId)
            applyStateIfSnapMatches(snapId: snapId, newState: .uploaded)
        } catch {
            Self.logger.warning(
                "snap upload failed (first attempt): \(error.localizedDescription, privacy: .private)"
            )
            try? await Task.sleep(for: retryDelay)
            do {
                try await repo.uploadProcessed(processed, snapId: snapId, userId: userId)
                applyStateIfSnapMatches(snapId: snapId, newState: .uploaded)
            } catch {
                Self.logger.error(
                    "snap upload failed (retry): \(error.localizedDescription, privacy: .private)"
                )
                applyStateIfSnapMatches(snapId: snapId, newState: .failed)
            }
        }
    }

    /// Mutates `formSnap`'s `.pending` state only if the current snap id
    /// still matches. Guards against the user removing the snap mid-upload.
    private func applyStateIfSnapMatches(snapId: UUID, newState: AttachedSnap.UploadState) {
        guard case .pending(var snap) = formSnap, snap.id == snapId else { return }
        snap.state = newState
        formSnap = .pending(snap)
    }

    /// Loads the bytes for a `PHPicker` selection. Bridges the callback-based
    /// `loadDataRepresentation` API into async/await.
    private static func loadData(from provider: NSItemProvider, uti: String) async throws -> Data {
        try await withCheckedThrowingContinuation { continuation in
            provider.loadDataRepresentation(forTypeIdentifier: uti) { data, error in
                if let error {
                    continuation.resume(throwing: error)
                } else if let data {
                    continuation.resume(returning: data)
                } else {
                    continuation.resume(throwing: URLError(.cannotDecodeContentData))
                }
            }
        }
    }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/features/pebblemedia/SnapUploadCoordinator.kt
package app.pbbls.android.features.pebblemedia

import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import app.pbbls.android.features.pebblemedia.models.AttachedSnap
import app.pbbls.android.features.pebblemedia.models.FormSnap
import app.pbbls.android.services.SnapWriteRepositing
import kotlinx.coroutines.delay
import java.util.UUID

private const val TAG = "snap-coordinator"

/**
 * Single source of truth for the photo attached to an in-progress pebble form —
 * ports iOS `SnapUploadCoordinator.swift` verbatim (design D3). Owns [formSnap]
 * (the form-layer state), [processedForRetry] (re-encoded bytes kept for a
 * user-tapped retry), and every transition: attach → upload (one auto-retry
 * after [retryDelayMillis], then FAILED) → retry → remove → cancel.
 *
 * Form-scoped, not app-level (design D6): Create/Edit screens construct one
 * with `remember` so an in-flight upload dies with the form. All Storage
 * side-effects are explicitly awaited — no sleep-based race in
 * [cancelAndCleanup]. [onLog] is injectable because the failure paths run in
 * JVM unit tests, where `android.util.Log` is unavailable (SnapURLCache rule).
 */
class SnapUploadCoordinator(
    private val repo: SnapWriteRepositing,
    initialSnap: FormSnap? = null,
    private val retryDelayMillis: Long = 2_000L,
    private val onLog: (String, Exception?) -> Unit = { message, e -> Log.w(TAG, message, e) },
) {
    var formSnap: FormSnap? by mutableStateOf(initialSnap)
        private set

    /**
     * Processed bytes kept around so a FAILED → user-tapped retry doesn't
     * re-decode + re-encode the pick. Cleared whenever [formSnap] clears.
     * Not UI-read, so a plain field.
     */
    var processedForRetry: ProcessedImage? = null
        private set

    /** True while the pending snap is still uploading. */
    val isUploading: Boolean
        get() = pendingState() == AttachedSnap.UploadState.UPLOADING

    /** True when the pending snap's last upload attempt failed. */
    val hasFailed: Boolean
        get() = pendingState() == AttachedSnap.UploadState.FAILED

    /** True while Save should be blocked (mid-upload, or failed and needing user action). */
    val isBlocking: Boolean
        get() = isUploading || hasFailed

    /**
     * The pending [AttachedSnap] only when fully uploaded — suitable for a
     * save payload. Null for none/existing/in-flight/failed (existing snaps
     * are encoded directly from [formSnap] by the payload builder).
     */
    fun pendingSnapForPayload(): AttachedSnap? =
        (formSnap as? FormSnap.Pending)
            ?.snap
            ?.takeIf { it.state == AttachedSnap.UploadState.UPLOADED }

    /**
     * Set the initial snap after construction — the edit screen has to wait
     * for its detail load before it knows whether the pebble has a snap.
     */
    fun seedExisting(snap: FormSnap?) {
        formSnap = snap
    }

    /** A processed pick enters the machine: pending-uploading, then [performUpload]. */
    suspend fun attach(
        processed: ProcessedImage,
        userId: String,
    ) {
        val snapId = UUID.randomUUID().toString().lowercase()
        formSnap =
            FormSnap.Pending(
                AttachedSnap(
                    id = snapId,
                    localThumb = processed.thumb,
                    state = AttachedSnap.UploadState.UPLOADING,
                ),
            )
        processedForRetry = processed
        performUpload(processed = processed, snapId = snapId, userId = userId)
    }

    /**
     * Re-run the upload after a user tap on the chip's retry button. Only
     * valid while pending with retained bytes; no-op otherwise.
     */
    suspend fun retryCurrent(userId: String) {
        val pending = formSnap as? FormSnap.Pending ?: return
        val processed = processedForRetry ?: return
        formSnap = FormSnap.Pending(pending.snap.copy(state = AttachedSnap.UploadState.UPLOADING))
        performUpload(processed = processed, snapId = pending.snap.id, userId = userId)
    }

    /**
     * Remove the current pending snap and fire the compensating Storage
     * delete. Safe in any upload state; no-op when not pending.
     */
    suspend fun removePending(userId: String) {
        val pending = formSnap as? FormSnap.Pending ?: return
        formSnap = null
        processedForRetry = null
        repo.deleteFiles(snapId = pending.snap.id, userId = userId)
    }

    /**
     * Remove the current existing snap: `delete_pebble_media` (throws to the
     * screen on failure — the form surfaces it), then fire-and-forget Storage
     * cleanup of the returned path. No-op when not existing.
     */
    suspend fun removeExisting() {
        val existing = formSnap as? FormSnap.Existing ?: return
        val storagePath = repo.deletePebbleMedia(snapId = existing.id)
        repo.deleteFiles(storagePrefix = storagePath)
        formSnap = null
    }

    /**
     * Cancel-button path: capture the pending id, clear state, and await the
     * compensating delete. Storage failures are logged inside the repo, never
     * thrown — Cancel must not be un-cancellable.
     */
    suspend fun cancelAndCleanup(userId: String) {
        val pending = formSnap as? FormSnap.Pending
        formSnap = null
        processedForRetry = null
        if (pending != null) {
            repo.deleteFiles(snapId = pending.snap.id, userId = userId)
        }
    }

    /**
     * Save (compose-pebble) failed with a pending snap attached: await the
     * compensating Storage delete but KEEP [formSnap] — the form stays open
     * so the user can retry. Use [cancelAndCleanup] to also dismiss.
     * (Verbatim iOS semantics — design D3.)
     */
    suspend fun handleSaveFailure(userId: String) {
        val pending = formSnap as? FormSnap.Pending ?: return
        repo.deleteFiles(snapId = pending.snap.id, userId = userId)
    }

    /**
     * Single upload path for [attach] and [retryCurrent]: on first failure,
     * wait [retryDelayMillis] and retry once; on second failure transition to
     * FAILED (bytes retained). State writes guard on snap-id match so a
     * removal mid-upload can't resurrect the chip.
     */
    private suspend fun performUpload(
        processed: ProcessedImage,
        snapId: String,
        userId: String,
    ) {
        try {
            repo.uploadProcessed(processed = processed, snapId = snapId, userId = userId)
            applyStateIfSnapMatches(snapId, AttachedSnap.UploadState.UPLOADED)
        } catch (first: Exception) {
            onLog("snap upload failed (first attempt)", first)
            delay(retryDelayMillis)
            try {
                repo.uploadProcessed(processed = processed, snapId = snapId, userId = userId)
                applyStateIfSnapMatches(snapId, AttachedSnap.UploadState.UPLOADED)
            } catch (second: Exception) {
                onLog("snap upload failed (retry)", second)
                applyStateIfSnapMatches(snapId, AttachedSnap.UploadState.FAILED)
            }
        }
    }

    private fun pendingState(): AttachedSnap.UploadState? = (formSnap as? FormSnap.Pending)?.snap?.state

    private fun applyStateIfSnapMatches(
        snapId: String,
        newState: AttachedSnap.UploadState,
    ) {
        val pending = formSnap as? FormSnap.Pending ?: return
        if (pending.snap.id != snapId) return
        formSnap = FormSnap.Pending(pending.snap.copy(state = newState))
    }
}
